presto-main/src/main/java/com/facebook/presto/operator/project/MergingPageOutput.java

```diff
@@ -0,0 +1,172 @@
+
+    private final List<Type> types;
+    private final PageBuilder pageBuilder;
+    private final Queue<PageProcessorOutput> inputQueue = new LinkedList<>();
+    private final Queue<Page> outputQueue = new LinkedList<>();
+
+    private final long minPageSizeInBytes;
+    private final long minRowCount;
+
+    private boolean finishing = false;
+
+    public MergingPageOutput(Iterable<? extends Type> types, int minPageSizeInBytes, int minRowCount)
+    {
+        this(types, minPageSizeInBytes, minRowCount, DEFAULT_MAX_PAGE_SIZE_IN_BYTES);
+    }
+
+    public MergingPageOutput(Iterable<? extends Type> types, int minPageSizeInBytes, int minRowCount, int maxPageSizeInBytes)
+    {
+        this.types = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+        this.minPageSizeInBytes = minPageSizeInBytes;
+        this.minRowCount = minRowCount;
+        pageBuilder = PageBuilder.withMaxPageSize(maxPageSizeInBytes, this.types);
+    }
+
+    public boolean needsInput()
+    {
+        return !finishing && outputQueue.isEmpty() && inputQueue.isEmpty();
+    }
+
+    public void addInput(PageProcessorOutput input)
+    {
+        requireNonNull(input, "input is null");
+        checkState(!finishing, "output is in finishing state");
+        inputQueue.add(input);
+    }
+
+    @Nullable
+    public Page getOutput()
+    {
+        if (!outputQueue.isEmpty()) {
+            return outputQueue.poll();
+        }
+
+        while (!inputQueue.isEmpty() && outputQueue.isEmpty()) {
+            Iterator<Page> input = inputQueue.peek();
+            if (input.hasNext()) {
+                process(input.next());
+            }
+            if (!input.hasNext()) {
+                inputQueue.poll();
+            }
+        }
+
+        if (inputQueue.isEmpty() && finishing) {
+            flush();
+        }
+
+        return outputQueue.poll();
+    }
+
+    public void finish()
+    {
+        finishing = true;
+    }
+
+    public boolean isFinished()
+    {
+        return finishing && inputQueue.isEmpty() && outputQueue.isEmpty() && pageBuilder.isEmpty();
+    }
+
+    private void process(Page page)
+    {
+        requireNonNull(page, "page is null");
+
+        // avoid memory copying for pages that are big enough
+        if (page.getSizeInBytes() >= minPageSizeInBytes || page.getPositionCount() >= minRowCount) {
+            flush();
+            outputQueue.add(page);
+            return;
+        }
+
+        buffer(page);
+    }
+
+    private void buffer(Page page)
+    {
+        for (int position = 0; position < page.getPositionCount(); position++) {
+            pageBuilder.declarePosition();
+            for (int channel = 0; channel < types.size(); channel++) {
+                Type type = types.get(channel);
+                type.appendTo(page.getBlock(channel), position, pageBuilder.getBlockBuilder(channel));
+            }
+        }
+        if (pageBuilder.isFull()) {
```

Comment:
This if statement needs to be inside the for loop over the positions in the page, or else you could get a page that's too big